### PR TITLE
feat: nlp example

### PR DESCRIPTION
**Pull request by theUnkownName**
WORK IN PROGRESS 

- [x] Flow is similar to Multires lyrics
- [x] At query time use Faiss with ref_indexer 
- [x] At query time use Annoy with ref_indexer 
- [x] At query time use Scann with ref_indexer 
- [x] Dummy random generator of text for input_fn.

### CODE REVIEW

File: stress_test_trigger/nlp_search/app.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+# for index
+def index():
+    with Flow.load_config('flows/index.yml') as index_flow:
+        index_flow.index(input_fn=document_generator(0, NUM_DOCS, NUM_CHUNKS), batch_size=BATCH_SIZE)
+
+
+# for search
+def query():
+    with Flow.load_config('flows/query.yml') as search_flow:
```
Comment (cristianmtr) [inline stress_test_trigger/nlp_search/app.py:74]: `query.yml` is empty?

Comment (theUnkownName) [reply stress_test_trigger/nlp_search/app.py:74]: Yeah for now, this is just for index at the moment

**Review by JoanFM — DISMISSED** [review]

File: stress-example/nlp_search/app.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+from jina.flow import Flow
+from jina import Document
+
+NUM_DOCS = 50
+NUM_CHUNKS = 5
+QUERY_NUM_DOCS = 1
+TOP_K = 3
+REQUEST_SIZE = 4
+IMG_HEIGHT = 224
```
Comment (cristianmtr) [inline stress-example/nlp_search/app.py:18]: Not used

File: stress-example/nlp_search/app.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+IMG_WIDTH = 224
```
Comment (cristianmtr) [inline stress-example/nlp_search/app.py:19]: Same

File: stress-example/nlp_search/app.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def config(indexer_query_type):
+    parallel = 1
```
Comment (cristianmtr) [inline stress-example/nlp_search/app.py:23]: But these are the same, right? So which one are you using when?

Comment (theUnkownName) [reply stress-example/nlp_search/app.py:23]: True, I'm not using the others, I'll clean it up

File: stress-example/nlp_search/flows/index.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  - name: encoder
+    show_exc_info: true
+    uses: pods/encoder.yml
+    parallel: $JINA_PARALLEL
+    timeout_ready: 600000
+    read_only: true
+  - name: chunk_idx
+    show_exc_info: true
+    uses: pods/chunk.yml
+    shards: $JINA_SHARDS
```
Comment (cristianmtr) [inline stress-example/nlp_search/flows/index.yml:17]: `shards` is the same as `parallel`. Maybe specify which ones are used when in the naming

`JINA_SHARDS_CHUNK_IDX` and `JINA_SHARDS_ENCODER` etc. And only use one of the names, either `shard` or `parallel`-based

**Review by cristianmtr — CHANGES_REQUESTED** [review]
Some clarifications

File: stress-example/nlp_search/app.py (hunk 4 above)
Comment (cristianmtr) [reply stress-example/nlp_search/app.py:23]: It seems they are both used. Line 26-27 you are setting the env. And then you are reading them in the yml

File: stress-example/nlp_search/pods/annoy_indexer.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
+!CompoundIndexer
+components:
+  - !AnnoyIndexer
+    with:
+      #index_filename: 'vec.gz' # Should be overwriten by ref_indexer index filename
```
Comment (cristianmtr) [inline stress-example/nlp_search/pods/annoy_indexer.yml:5]: Then should this be left here?

File: stress-example/nlp_search/pods/faiss_indexer.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
+!CompoundIndexer
+components:
+  - !FaissIndexer
+    with:
+      index_key: 'Flat' #TODO: Find suitable index
+      train_filepath: 'None'
+      #index_filename: 'vec.gz' # Should be overwriten by ref_indexer index filename
```
Comment (cristianmtr) [inline stress-example/nlp_search/pods/faiss_indexer.yml:7]: clean up here too?

File: stress-example/nlp_search/pods/faiss_indexer.yml [hunk 8]
```diff
@@ -0,0 +1,50 @@
+!CompoundIndexer
+components:
+  - !FaissIndexer
+    with:
+      index_key: 'Flat' #TODO: Find suitable index
```
Comment (cristianmtr) [inline stress-example/nlp_search/pods/faiss_indexer.yml:5]: Is TODO still relevant?

File: stress-example/nlp_search/pods/scann_indexer.yml [hunk 9]
```diff
@@ -0,0 +1,49 @@
+!CompoundIndexer
+components:
+  - !ScannIndexer
+    with:
+      distance_measure: 'squared_l2'
+      #index_filename: 'vec.gz' # Should be overwriten by ref_indexer index filename
```
Comment (cristianmtr) [inline stress-example/nlp_search/pods/scann_indexer.yml:6]: same

File: stress-example/nlp_search/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,5 @@
+#jina[hub,http]==0.9.17
```
Comment (cristianmtr) [inline stress-example/nlp_search/requirements.txt:1]: Why is `jina` commented out?

File: stress-example/nlp_search/pods/annoy_indexer.yml (hunk 6 above)
Comment (JoanFM) [reply stress-example/nlp_search/pods/annoy_indexer.yml:5]: good for clarification

File: stress-example/nlp_search/pods/faiss_indexer.yml (hunk 8 above)
Comment (JoanFM) [reply stress-example/nlp_search/pods/faiss_indexer.yml:5]: yes, it is

File: stress-example/nlp_search/pods/annoy_indexer.yml (hunk 6 above)
Comment (theUnkownName) [reply stress-example/nlp_search/pods/annoy_indexer.yml:5]: I thought it was good to left as to clarify why I'm not using that name

File: stress-example/nlp_search/pods/faiss_indexer.yml (hunk 7 above)
Comment (theUnkownName) [reply stress-example/nlp_search/pods/faiss_indexer.yml:7]: Same

File: stress-example/nlp_search/pods/faiss_indexer.yml (hunk 8 above)
Comment (theUnkownName) [reply stress-example/nlp_search/pods/faiss_indexer.yml:5]: It's waiting for the Faiss work

File: stress-example/nlp_search/requirements.txt (hunk 10 above)
Comment (theUnkownName) [reply stress-example/nlp_search/requirements.txt]: True, I got it back

**Review by JoanFM — APPROVED** [review]

File: stress-example/image_search/flows/query.yml [hunk 11]
```diff
@@ -17,7 +17,7 @@ pods:
     polling: all
     uses_after: _merge_matches
     timeout_ready: 100000 # larger timeout as in query time will read all the data
-    volumes: './workspace:/workspace/workspace'
+    volumes: './workspace:/docker_workspace/workspace'
```
Comment (JoanFM) [inline stress-example/image_search/flows/query.yml:20]: can u try simply `/docker-wokspace` without the extra `workspace`?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: stress-example/image_search/pods/annoy_indexer.yml [hunk 12]
```diff
@@ -8,24 +8,24 @@ components:
...
     metas:
       name: docidx
-      workspace: '/workspace/workspace'
+      workspace: '/docker_workspace/workspace'
 metas:
   name: vec_idx
   py_modules:
-    - workspace/__init__.py
-  workspace: '/workspace/workspace'
+    - docker_workspace/__init__.py
```
Comment (JoanFM) [inline stress-example/image_search/pods/annoy_indexer.yml:27]: this `py_modules` should not be changed

File: stress-example/image_search/pods/faiss_indexer.yml [hunk 13]
```diff
@@ -8,24 +8,24 @@ components:
...
     metas:
       name: docidx
-      workspace: '/workspace/workspace'
+      workspace: '/docker_workspace/workspace'
 metas:
   name: vec_idx
   py_modules:
-    - workspace/__init__.py
-  workspace: '/workspace/workspace'
+    - docker_workspace/__init__.py
```
Comment (JoanFM) [inline stress-example/image_search/pods/faiss_indexer.yml:27]: neither here

File: stress-example/image_search/pods/scann_indexer.yml [hunk 14]
```diff
@@ -7,24 +7,24 @@ components:
...
     metas:
       name: docidx
-      workspace: '/workspace/workspace'
+      workspace: '/docker_workspace/workspace'
 metas:
   name: vec_idx
   py_modules:
-    - workspace/__init__.py
-  workspace: '/workspace/workspace'
+    - docker_workspace/__init__.py
```
Comment (JoanFM) [inline stress-example/image_search/pods/scann_indexer.yml:26]: or here

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: stress-example/image_search/pods/annoy_indexer.yml (hunk 12 above)
Comment (theUnkownName) [reply stress-example/image_search/pods/annoy_indexer.yml:27]: If it's not changed it doesn't find the correct one

File: stress-example/image_search/flows/query.yml (hunk 11 above)
Comment (theUnkownName) [reply stress-example/image_search/flows/query.yml:20]: But how is this going to work? how is it going to find it? If I try it like that doesn't get any data:
`vec_idx/2/ContainerRuntime@34340[I]:num_dim extracted from 'ref_indexer' to None`

File: stress-example/image_search/app.py [hunk 15]
```diff
@@ -36,11 +36,11 @@ def random_docs(start, end):
...
 def config(indexer_query_type):
-    parallel = 1
-    shards = 2
+    shards_encoder = 1
+    shards_indexers = 2
 
-    os.environ['JINA_PARALLEL'] = str(parallel)
-    os.environ['JINA_SHARDS'] = str(shards)
+    os.environ['JINA_SHARDS_ENCODER'] = str(shards_encoder)
+    os.environ['JINA_SHARDS_INDEXER'] = str(shards_indexers)
```
Comment (cristianmtr) [inline stress-example/image_search/app.py:43]: `INDEXERS`, to make it clear it's used for both (and to make it consistent with my changes)

**Review by cristianmtr — CHANGES_REQUESTED** [review]
`INDEXERS`

File: stress-example/image_search/app.py (hunk 15 above)
Comment (theUnkownName) [reply stress-example/image_search/app.py]: Done

**Review by JoanFM — APPROVED** [review]

**Review by JoanFM — CHANGES_REQUESTED** [review]
Do not assert, we will try just to log when this happens

**Comment by cristianmtr** [comment]
> Do not assert, we will try just to log when this happens

I think we should assert for now, until we figure out the issue with workspaces etc. I might've missed it several times because the flow runs fine, but there are no matches returned.

**Comment by JoanFM** [comment]
> > Do not assert, we will try just to log when this happens
> 
> I think we should assert for now, until we figure out the issue with workspaces etc. I might've missed it several times because the flow runs fine, but there are no matches returned.

Ok

**Review by JoanFM — APPROVED** [review]

File: stress-example/nlp_search/flows/index.yml [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+version: '1'
+pods:
+  - name: segmenter
+    show_exc_info: true
+    uses: pods/segment.yml
+    read_only: true
+  - name: encoder
+    show_exc_info: true
+    uses: pods/encoder.yml
+    parallel: $JINA_SHARDS_ENCODER
```
Comment (cristianmtr) [inline stress-example/nlp_search/flows/index.yml:11]:
```suggestion
    shards: $JINA_SHARDS_ENCODER
```

Make sure you've replaced all usage of `parallel` with `shards`. It's the same, but it's a bit confusing to use both (we should just stick to one for consistency)

File: stress-example/image_search/flows/index.yml [hunk 17]
```diff
@@ -4,17 +4,17 @@ with:
 pods:
   encoder:
     uses: pods/encoder.yml
-    parallel: $JINA_PARALLEL
+    parallel: $JINA_SHARDS_ENCODER
```
Comment (cristianmtr) [inline stress-example/image_search/flows/index.yml:7]:
```suggestion
    shards: $JINA_SHARDS_ENCODER
```

File: stress-example/image_search/flows/query.yml [hunk 18]
```diff
@@ -7,20 +7,20 @@ with:
 pods:
   encoder:
     uses: pods/encoder.yml
-    parallel: $JINA_PARALLEL
+    parallel: $JINA_SHARDS_ENCODER
```
Comment (cristianmtr) [inline stress-example/image_search/flows/query.yml:10]: Here too

File: stress-example/nlp_search/flows/query.yml [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+version: '1'
+with:
+  read_only: true  # better add this in the query time
+  rest_api: false
+  port_expose: $JINA_PORT
+pods:
+  - name: chunk_seg
+    show_exc_info: true
+    uses: pods/segment.yml
+    parallel: $JINA_SHARDS_ENCODER
```
Comment (cristianmtr) [inline stress-example/nlp_search/flows/query.yml:11]: Here too

File: stress-example/nlp_search/flows/query.yml [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+  - name: encoder
+    show_exc_info: true
+    uses: pods/encoder.yml
+    parallel: $JINA_SHARDS_ENCODER
```
Comment (cristianmtr) [inline stress-example/nlp_search/flows/query.yml:15]: Here too

Comment (theUnkownName) [reply stress-example/nlp_search/flows/query.yml]: Done

**Review by cristianmtr — APPROVED** [review]
looks good, great job :+1:

**Pull request merged**